src/main/scala/com/softwaremill/realworld/articles/ArticlesRepository.scala

```diff
@@ -87,6 +87,17 @@ class ArticlesRepository(quill: Quill.Sqlite[SnakeCase]):
         case None    => ZIO.fail(Exceptions.NotFound(s"Article with slug $slug doesn't exist."))
       }
 
+  def findArticleBySlug(slug: String): Task[ArticleRow] =
+    run(
+      queryArticle
+        .filter(a => a.slug == lift(slug))
+    )
+      .map(_.headOption)
+      .flatMap {
```

Comment:
this is third place with same flatMap (2 here, 1 in service)
```
maybe we could add function like:
  def handleArticleProcessingResult[T](option: Option[T], slug: String): Task[T] = // or handleArticleNotFoundError or something similar
    option match {
      case Some(value) => ZIO.succeed(value)
      case None        => ZIO.fail(Exceptions.NotFound(s"Article with slug $slug doesn't exist."))
    }
```
and re-use it:
`.flatMap(handleArticleProcessingResult(_, slug))`

Reply:
or even 4th if we count
`findBySlugAsSeenBy(articleId: Int, email: String)`
and we would implement this function more generic like:
`.flatMap(handleArticleProcessingResult(_, s"error message here"))
...
def handleArticleProcessingResult[T](option: Option[T], errorMessage: String): Task[T] = // or handleArticleNotFoundError or something similar
    option match {
      case Some(value) => ZIO.succeed(value)
      case None        => ZIO.fail(Exceptions.NotFound(errorMessage))
    }
`

Reply:
IMO repository is not the best place for that regarding code readability/maintenance and such method should be placed into separate file, for example we could name it ArticleUtils.
Besides this LGTM ;)

Reply:
It's used by the service and repository.

Reply:
Ah :) Though this is a bit suspicious, it looks more live service-layer code

Reply:
Then we could handle it on service side in both cases - it would be more proper and as the benefit we could make _handleArticleProcessingResult_ just a private function. However this would mean more repeated code in several service's functions but I think it's worth it. And what do you think?
BTW there is similar issue with _findComment_ function - maybe it could be changed on the occasion as there are just 2 occurrences of the usage and it should be simple to change it.

Reply:
@FawnPug I added handleProcessingResult as a private function in service and I use it to service findArticle..., findComment and findBySlugAsSeenBy methods from article service and repository. Is it ok or should it look different?

Reply:
@wydra98 now when repository functions return optional I can see a better solution :). It's much better to use option's api to handle this error than adding custom function.
[I've just created PR with my proposal.](https://github.com/softwaremill/realworld-tapir-zio/pull/80)
I think this is a simpler and clearer solution.

Reply:
(please don't wait for me and feel free to approve/merge [it](https://github.com/softwaremill/realworld-tapir-zio/pull/80) to your branch if you like it)

Reply:
Yes, i like this code so much, so I'm going to approve and merge it to my branch ;)